### ISSUE TITLE
Feature request: rANS entropy-coded quantization for 1.84x compression over 4-bit

### ISSUE THREAD

**Issue by dreddyamaze** [ml-explore/mlx#3042]
### 🚀 Feature request: Add **rANS entropy-coded quantization** for 1.84x compression over 4-bit

**TL;DR**  
4-bit quantized weights have measured entropy of **~2.17 bits/weight**, not 4.0. By applying rANS entropy coding on top of 4-bit quantization, we achieve **1.84x additional lossless compression**. A novel V2 per-row encoding scheme enables efficient fused GPU decode with only 1.1-1.5x overhead vs pre-decoded weights.

---

## 1 Background & motivation

- **Entropy gap**: 4-bit quantized weights cluster around a few values. Measured entropy on Llama/Qwen models is ~2.17 bits vs 4.0 bits stored.
- **Bandwidth bound**: LLM inference on Apple Silicon is memory-bound; reducing weight traffic directly improves tok/s.
- **Lossless**: Unlike lower-bit quantization, entropy coding preserves exact 4-bit values—no additional accuracy loss.

| Model | 4-bit Size | ECQ Size | Reduction |
|-------|-----------|----------|-----------|
| 7B | 3.5 GB | ~1.9 GB | 1.84x |
| 13B | 6.5 GB | ~3.5 GB | 1.84x |
| 70B | 35 GB | ~19 GB | 1.84x |

---

## 2 Key innovation: V2 per-row encoding

rANS is inherently sequential—decoding symbol N requires the state from decoding symbols 0→N-1. A naive flat encoding forces each GPU threadgroup to decode the *entire* weight matrix.

**V2 solution**: Encode each output row independently. Each threadgroup decodes only its row's compressed stream.

```
V1: O(rows × total_symbols) decode work  ❌
V2: O(total_symbols) decode work         ✅
```

| Layer Size | V1 | V2 | Speedup |
|------------|-----|-----|---------|
| 256×512 | 9.87ms | 0.33ms | **30x** |
| 512×1024 | 74.3ms | 0.50ms | **148x** |
| 1024×2048 | 615ms | 1.08ms | **571x** |

---

## 3 Decode strategies implemented

| Mode | Memory Footprint | Inference Overhead | Best For |
|------|------------------|-------------------|----------|
| **FUSED_V2** (default) | 1.84x smaller | 1.1-1.5x | Memory-constrained, large models |
| **CACHED** | Same as 4-bit | 1.0x | Speed-critical, sufficient RAM |
| **GPU_ASYNC** | 1.84x smaller | ~1.0x | Very deep models (50+ layers) |

---

## 4 Proposed API

```python
import mlx.nn as nn
from mlx.nn.layers.entropy_coded import EntropyCodedLinear, entropy_quantize

# Option 1: Convert single layer
linear = nn.Linear(4096, 4096)
ec_layer = EntropyCodedLinear.from_linear(linear, decode_mode="fused_v2")

# Option 2: Convert entire model
model = MyModel()
entropy_quantize(model, decode_mode="fused_v2")

# Inference (1.84x smaller weights, ~1.2x overhead)
y = ec_layer(x)
```

---

## 5 Implementation status

We have a working implementation with:

| Component | Status | Lines |
|-----------|--------|-------|
| Metal kernel V1 (fused decode+GEMV) | ✅ | ~390 |
| Metal kernel V2 (per-row) | ✅ | ~256 |
| Metal async decode | ✅ | ~250 |
| C++ `EntropyCodedMatmul` primitive | ✅ | |
| C++ `EntropyCodedMatmulV2` primitive | ✅ | |
| C++ `EntropyDecodeAsync` primitive | ✅ | |
| Python `EntropyCodedLinear` layer | ✅ | ~760 |
| Python bindings | ✅ | |

**Files changed vs origin/main:**
```
mlx/backend/cpu/quantized.cpp             |  24 +++
mlx/backend/metal/kernels/entropy_coded.h       (new)
mlx/backend/metal/kernels/entropy_coded_v2.h    (new)
mlx/backend/metal/kernels/entropy_decode_async.h (new)
mlx/backend/metal/quantized.cpp           | 184 +++
mlx/ops.cpp                               | 174 +++
mlx/ops.h                                 |  50 +++
mlx/primitives.cpp                        |  49 +++
mlx/primitives.h                          | 102 +++
python/mlx/nn/layers/__init__.py          |   5 +
python/mlx/nn/layers/entropy_coded.py           (new)
python/src/ops.cpp                        | 133 +++
```

---

## 6 Benchmarks

**Synthetic layers (M3 Pro):**

| Layer | FUSED_V2 | CACHED | Overhead |
|-------|----------|--------|----------|
| 1024×2048 | 1.08ms | 0.92ms | 1.17x |
| 2048×4096 | 2.31ms | 1.98ms | 1.17x |
| 4096×4096 | 4.12ms | 3.54ms | 1.16x |

**Real model (SmolLM-135M):**
- FUSED_V2 has higher overhead (~7x) on small models due to kernel launch overhead dominating
- For 7B+ models where inference is memory-bound, the 1.84x bandwidth reduction translates to significant speedup

---

## 7 References

1. [Asymmetric Numeral Systems (rANS)](https://arxiv.org/abs/1402.3392) - Duda, 2014
2. [Interleaved entropy coders](https://fgiesen.wordpress.com/2014/02/18/rans-with-static-probability-distributions/) - Giesen, 2014

---

**Questions for maintainers:**
1. Is this approach of interest for integration into MLX/mlx-lm?
2. Would you prefer this as a separate experimental module or integrated into core quantization?
3. Any feedback on the Metal kernel design (fused decode in GEMV)?

Happy to open a PR with the full implementation.
